File: cpp/src/render/utils/vega/vega_choropleth_map/vega_choropleth_map.cpp
/*
 * Copyright (C) 2019-2020 Zilliz. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <string>
#include <utility>

#include "render/utils/vega/vega_choropleth_map/vega_choropleth_map.h"

namespace arctern {
namespace render {

VegaChoroplethMap::VegaChoroplethMap(const std::string& json) { Parse(json); }

std::string VegaChoroplethMap::Build() {
  // TODO: add Build() api to build a vega json string.
  return "";
}

void VegaChoroplethMap::Parse(const std::string& json) {
  rapidjson::Document document;
  document.Parse(json.c_str());

  if (document.Parse(json.c_str()).HasParseError()) {
    printf("json format error\n");
    return;
  }

  if (!JsonLabelCheck(document, "width") || !JsonLabelCheck(document, "height") ||
      !JsonTypeCheck(document["width"], rapidjson::Type::kNumberType) ||
      !JsonTypeCheck(document["height"], rapidjson::Type::kNumberType)) {
    return;
  }
  window_params_.mutable_width() = document["width"].GetInt();
  window_params_.mutable_height() = document["height"].GetInt();

  if (!JsonLabelCheck(document, "marks") ||
      !JsonTypeCheck(document["marks"], rapidjson::Type::kArrayType) ||
      !JsonSizeCheck(document["marks"], "marks", 1) ||
      !JsonLabelCheck(document["marks"][0], "encode") ||
      !JsonLabelCheck(document["marks"][0]["encode"], "enter")) {
    return;
  }
  rapidjson::Value mark_enter;
  mark_enter = document["marks"][0]["encode"]["enter"];

  // parse bounding box
  if (!JsonLabelCheck(mark_enter, "bounding_box") ||
      !JsonLabelCheck(mark_enter["bounding_box"], "value") ||
      !JsonTypeCheck(mark_enter["bounding_box"]["value"], rapidjson::Type::kArrayType) ||
      !JsonSizeCheck(mark_enter["bounding_box"]["value"], "bounding_box.value", 4)) {
    return;
  }
  for (int i = 0; i < 4; i++) {
    if (!JsonTypeCheck(mark_enter["bounding_box"]["value"][i],
                       rapidjson::Type::kNumberType)) {
      return;
    }
  }
  bounding_box_.longitude_left = mark_enter["bounding_box"]["value"][0].GetDouble();
  bounding_box_.latitude_left = mark_enter["bounding_box"]["value"][1].GetDouble();
  bounding_box_.longitude_right = mark_enter["bounding_box"]["value"][2].GetDouble();
  bounding_box_.latitude_right = mark_enter["bounding_box"]["value"][3].GetDouble();

  // parse color style
  if (!JsonLabelCheck(mark_enter, "color_style") ||
      !JsonLabelCheck(mark_enter["color_style"], "value") ||
      !JsonTypeCheck(mark_enter["color_style"]["value"], rapidjson::Type::kStringType)) {
    return;
  }
  auto color_style_string = std::string(mark_enter["color_style"]["value"].GetString());
  if (color_style_string == "blue_to_red") {
    color_style_ = ColorStyle::kBlueToRed;
  } else if (color_style_string == "skyblue_to_white") {
    color_style_ = ColorStyle::kSkyBlueToWhite;
  } else if (color_style_string == "purple_to_yellow") {
    color_style_ = ColorStyle::kPurpleToYellow;
  } else if (color_style_string == "red_transparency") {
    color_style_ = ColorStyle::kRedTransParency;
  } else if (color_style_string == "blue_transparency") {
    color_style_ = ColorStyle::kBlueTransParency;
  } else if (color_style_string == "blue_green_yellow") {
    color_style_ = ColorStyle::kBlueGreenYellow;
  } else if (color_style_string == "white_blue") {
    color_style_ = ColorStyle::kWhiteToBlue;
  } else if (color_style_string == "blue_white_red") {
    color_style_ = ColorStyle::kBlueWhiteRed;
  } else if (color_style_string == "green_yellow_red") {
    color_style_ = ColorStyle::kGreenYellowRed;
  } else {
    std::string msg = "unsupported color style '" + color_style_string + "'.";
    // TODO: add log here
  }

  // parse ruler
  if (!JsonLabelCheck(mark_enter, "ruler") ||
      !JsonLabelCheck(mark_enter["ruler"], "value") ||
      !JsonTypeCheck(mark_enter["ruler"]["value"], rapidjson::Type::kArrayType) ||
      !JsonSizeCheck(mark_enter["ruler"]["value"], "ruler.value", 2)) {
    return;
  }
  for (int i = 0; i < 2; i++) {
    if (!JsonTypeCheck(mark_enter["ruler"]["value"][i], rapidjson::Type::kNumberType)) {
      return;
    }
  }
  ruler_ = std::make_pair(mark_enter["ruler"]["value"][0].GetDouble(),
                          mark_enter["ruler"]["value"][1].GetDouble());

  // parse opacity
  if (!JsonLabelCheck(mark_enter, "opacity") ||
      !JsonLabelCheck(mark_enter["opacity"], "value") ||
      !JsonTypeCheck(mark_enter["opacity"]["value"], rapidjson::Type::kNumberType)) {
    return;
  }
  opacity_ = mark_enter["opacity"]["value"].GetDouble();
}

}  // namespace render
}  // namespace arctern

File: python/arctern/util/vega/choroplethmap/vega_choroplethmap.py
# Copyright (C) 2019-2020 Zilliz. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from arctern.util.vega.vega_node import (Width, Height, Description, Data,
                                            Scales, RootMarks, Root)


class Marks(RootMarks):
    """
        Top-Level Vega Specification Property: Marks
    """

    class Encode:
        class Value:
            def __init__(self, v: int or float or str):
                self.v = v

            def to_dict(self):
                dic = {
                    "value": self.v
                }
                return dic

        def __init__(self, bounding_box_min: Value, bounding_box_max: Value,
                     color_style: Value, ruler: Value, opacity: Value, coordinate_system: Value):
            if not (isinstance(bounding_box_min.v, str)
                    and isinstance(bounding_box_max.v, str)
                    and isinstance(color_style.v, str)
                    and isinstance(ruler.v, list)
                    and isinstance(opacity.v, float)
                    and isinstance(coordinate_system.v, str)):
                # TODO error log here
                assert 0, "illegal"
            self._bounding_box_min = bounding_box_min
            self._bounding_box_max = bounding_box_max
            self._color_style = color_style
            self._ruler = ruler
            self._opacity = opacity
            self._coordinate_system = coordinate_system

        def to_dict(self):
            dic = {
                "enter": {
                    "bounding_box_min": self._bounding_box_min.to_dict(),
                    "bounding_box_max": self._bounding_box_max.to_dict(),
                    "color_style": self._color_style.to_dict(),
                    "ruler": self._ruler.to_dict(),
                    "opacity": self._opacity.to_dict(),
                    "coordinate_system": self._coordinate_system.to_dict()
                }
            }
            return dic

    def __init__(self, encode: Encode):
        self.encode = encode

    def to_dict(self):
        dic = [{
            "encode": self.encode.to_dict()
        }]
        return dic

class VegaChoroplethMap:
    def __init__(self, width: int, height: int, bounding_box_min: str, bounding_box_max: str,
                 color_style: str, ruler: list, opacity: float, coordinate_system: str):
        self._width = width
        self._height = height
        self._bounding_box_min = bounding_box_min
        self._bounding_box_max = bounding_box_max
        self._color_style = color_style
        self._ruler = ruler
        self._opacity = opacity
        self._coordinate_system = coordinate_system

    def build(self):
        description = Description(desc="building_weighted_2d")
        data = Data(name="data", url="/data/data.csv")
        domain = Scales.Scale.Domain("data", "c0")
        scale = Scales.Scale("building", "linear", domain)
        scales = Scales([scale])
        encode = Marks.Encode(bounding_box_min=Marks.Encode.Value(self._bounding_box_min),
                              bounding_box_max=Marks.Encode.Value(self._bounding_box_max),
                              color_style=Marks.Encode.Value(self._color_style),
                              ruler=Marks.Encode.Value(self._ruler),
                              opacity=Marks.Encode.Value(self._opacity),
                              coordinate_system=Marks.Encode.Value(self._coordinate_system))
        marks = Marks(encode)
        root = Root(Width(self._width), Height(self._height), description,
                    data, scales, marks)

        root_json = json.dumps(root.to_dict(), indent=2)
        return root_json
    def coor(self):
        return self._coordinate_system
    
    def bounding_box_min(self):
        return self._bounding_box_min
    
    def bounding_box_max(self):
        return self._bounding_box_max

    def height(self):
        return self._height

    def width(self):
        return self._width
